libhpcalcs/src/prime_cmd.c

```diff
@@ -360,47 +364,124 @@ HPEXPORT int HPCALL calc_prime_r_recv_screen(calc_handle * handle, calc_screensh
+        return res;
+    }
+    return res;
+}
+
+HPEXPORT int HPCALL calc_prime_r_disable_new_protocol(calc_handle * handle) {
+    int res;
+    if (handle != NULL) {
+        // While the prime is in thew new protocol mode, it sends us packages
+        // starting with 0xFE. We need to eliminate those now (and also wait for
+        // the status response).
+        res = calc_prime_r_check_ready(handle, NULL, NULL);
+        handle->protocol_version = 0;
+    }
+    else {
+        res = ERR_INVALID_HANDLE;
+        hpcalcs_error("%s: handle is NULL", __FUNCTION__);
+    }
+    return res;
+}
+
 HPEXPORT int HPCALL calc_prime_s_send_file(calc_handle * handle, files_var_entry * file) {
     int res;
     if (handle != NULL && file != NULL) {
+        uint32_t offset = 0;
+        uint32_t header_size = 8;
         uint8_t namelen = (uint8_t)char16_strlen(file->name) * 2;
-        uint32_t size = 10 - 6 + namelen + file->size; // Size of the data after the header.
-        prime_vtl_pkt * pkt = prime_vtl_pkt_new(size + 6); // Add size of the header.
-        hpcalcs_debug("Virtual packet has size %" PRIu32 " (%" PRIx32 ")\n", size, size);
+        uint32_t size = namelen + file->size + 12; // Size of the data plus something
+        uint32_t other_size = namelen + file->size + 6; // Also size of the data plus something
+        prime_vtl_pkt * pkt;
+
+        // Some text editors add the UTF-16LE BOM at the beginning of the file, but the SDKV0.30 firmware version chokes on it.
+        // Therefore, skip the BOM.
+        if (   (file->type == PRIME_TYPE_PRGM || file->type == PRIME_TYPE_NOTE)
+            && (file->data[0] == 0xFF && file->data[1] == 0xFE)
+           ) {
+            offset = 2;
+            size -= 2;
+            other_size -= 2;
+        }
+
+        pkt = prime_vtl_pkt_new(size + header_size); // Add size of the header.
         if (pkt != NULL) {
             uint8_t * ptr;
             uint16_t crc16;
-            uint32_t offset = 0;
-
-            // Some text editors add the UTF-16LE BOM at the beginning of the file, but the SDKV0.30 firmware version chokes on it.
-            // Therefore, skip the BOM.
-            if (   (file->type == PRIME_TYPE_PRGM || file->type == PRIME_TYPE_NOTE)
-                && (file->data[0] == 0xFF && file->data[1] == 0xFE)
-               ) {
-                offset = 2;
-                size -= 2;
-            }
 
             pkt->cmd = CMD_PRIME_RECV_FILE;
             ptr = pkt->data;
-            *ptr++ = CMD_PRIME_RECV_FILE;
+
+            // Command sequence. Connectivity kit increments this after each command, but the Prime seems to ignore it.
             *ptr++ = 0x01;
-            *ptr++ = (uint8_t)((size >> 24) & 0xFF);
-            *ptr++ = (uint8_t)((size >> 16) & 0xFF);
-            *ptr++ = (uint8_t)((size >>  8) & 0xFF);
+            *ptr++ = 0x00;
+            *ptr++ = 0x00;
+            *ptr++ = 0x00;  
+
             *ptr++ = (uint8_t)((size      ) & 0xFF);
```

Comment:
Yeah, this size is little endian. The other one below is big endian. No idea why.